src/awscli-asset.ts

```diff
@@ -0,0 +1,18 @@
+import * as path from 'path';
+import { FileSystem } from 'aws-cdk-lib';
+import * as s3_assets from 'aws-cdk-lib/aws-s3-assets';
+import { Construct } from 'constructs';
+
+/**
+ * A CDK Asset construct that contains the AWS CLI.
+ */
+export class AwsCliAsset extends s3_assets.Asset {
```

Comment:
I like this being an asset, but i'm reluctant to do so at the expense of a somewhat awkward API as proposed in https://github.com/aws/aws-cdk/pull/21933, especially since the use-case we are trying to address is creating a `lambda.LayerVersion` object from an asset, not a `lambda.Code` object.

I realize we'll have to create a `lambda.Code` object from an asset at some point, but I believe you can utilize the `lambda.Code.fromBucket` function for that. 

This gives us 3 options as I see it.

#### 1) Provide a conversion method in the asset

```ts
export class AwsCliAsset extends s3_assets.Asset {
  
  ...
  ...

  public toLambdaLayerVersion(options: lambda.LayerVersionOptions = {}): lambda.LayerVersion {
    return new lambda.LayerVersion(this, 'LambdaLayerVersion', {
      code: lambda.Code.fromBucket(this.bucket, this.s3ObjectKey),
      ...options,
    });
  }
}
```

```ts
const awscli = new AwsCliAsset(...);
const layer = awscli.toLambdaLayerVersion({ description: '/opt/awscli/aws' });
```

This is fairly simple, and further justifies the fact `AwsCliAsset` extends `Asset`, since it now provide extended capabilities of turning an asset into a layer. Only downside is that its not very idiomatic to CDK to use `toXXX` methods. 

#### 2) Provide a new factory function for creating layers

```ts
export class LayerVersion extends LayerVersionBase {

  ...
  ...

  public static fromAsset(scope: Construct, id: string, asset: s3_assets.Asset, options: LayerVersionOptions) {
    return new LayerVersion(scope, id, {
      code: Code.fromBucket(asset.bucket, asset.s3ObjectKey),
      ...options,
    });
  }
}
```

```ts
const awscli = new AwsCliAsset(...);
const layer = lambda.LayerVersion.fromAsset(this, 'AwsCliLayer', { description: '/opt/awscli/aws' });
```

This seems legit. Only small issue I have with it is adding an API to `aws-cdk-lib` that isn't strictly required.

#### 3) Use existing API to create the layer

```ts
const awscli = new AwsCliAsset(...);
const layer = new lambda.LayerVersion(stack, 'AwsCliLayer', {
  code: lambda.Code.fromBucket(awscli.bucket, awscli.s3ObjectKey);
  description: '/opt/awscli/aws',
});
```

This puts a non trivial burden on the caller, so not great; but also not horrible because we're not expecting much third-party usage of this library. Its mainly for us at this point. And if we see it's getting usage we can provide a simpler API like option 1 or 2. 

-----------------------

WDYT?

Reply:
Option 2 is my favorite. I think that's better than option 1, since we will only write the "LayerVersion from Asset" logic one time in aws-cdk-lib. 

Option 3 is also a possibility, and less work upfront. I want to make sure I understand what you are proposing here. That option would mean no change to this PR, and AwsCliLayer in aws-cdk-lib would look like this, right?

```typescript
export class AwsCliLayer extends lambda.LayerVersion {
  private static assetPackage: any;

  constructor(scope: Construct, id: string) {
    AwsCliLayer.assetPackage = ... ; // Dynamic require logic goes here...
    const awscli = new AwsCliLayer.assetPackage.AwsCliAsset(scope, `${id}-asset`);

    super(scope, id, {
      code: lambda.Code.fromBucket(awscli.bucket, awscli.s3ObjectKey),
      description: '/opt/awscli/aws',
    });
  }
}
```

Reply:
Yeap that's what I mean. Actually option 3 is the only option that lends itself to an `AwsCliLayer` class that extends `lambda.LayerVersion` in `aws-cdk-lib`. Option 2 is just a convenience function to create a layer from an asset, which we could provide as well, but I don't see a reason for it just now.